File: src/vault/provider.ts
// The public vault-plugin contract (spec: "Modular vaults"). v2.0 ships one
// provider (menv-local); adding another must require zero changes outside the
// provider module and the registry map in vault/registry.ts.

// Resolved auth material for one vault. Provider-interpreted: for menv-local
// `secret` is the encryption passphrase; a future remote provider would read
// its token from here.
export interface VaultAuth {
  secret?: string;
}

export interface VaultInitContext {
  // Repo root — providers resolve repo-relative paths (e.g. menv-local's
  // vaultConfig.filename) against this.
  root: string;
  auth: VaultAuth;
}

export interface VaultSession {
  get(key: string): Promise<string | undefined>;
  set(key: string, value: string): Promise<void>;
  remove(key: string): Promise<void>;
  list(): Promise<string[]>;
  close(): Promise<void>;
}

export interface VaultProvider {
  readonly type: string;
  init(config: unknown, ctx: VaultInitContext): Promise<VaultSession>;
}

File: go/internal/vault/provider.go
package vault

import (
	"encoding/json"

	"github.com/nikrabaev/menv/internal/core"
)

// VaultAuth is the resolved auth material for one vault. Interpretation is
// provider-specific: menv-local uses secret as the age passphrase.
type VaultAuth struct {
	Secret string // empty means no auth supplied
	HasSecret bool // distinguishes "no auth" from empty-string passphrase
}

// VaultInitContext is passed to VaultProvider.Init.
type VaultInitContext struct {
	Root string    // repo root — providers resolve relative paths against this
	Auth VaultAuth
}

// VaultProvider is the plugin contract for vault backends.
type VaultProvider interface {
	Type() string
	// Init parses config (provider-specific JSON) and returns a ready session.
	// config is json.RawMessage so each provider controls its own schema.
	Init(config json.RawMessage, ctx VaultInitContext) (core.VaultSession, error)
}
